mlir/include/mlir/IR/BuiltinTypes.h

```diff
@@ -277,7 +277,7 @@ class RankedTensorType::Builder {
     if (storage.empty())
       storage.append(shape.begin(), shape.end());
     storage.erase(storage.begin() + pos);
-    shape = {storage.data(), storage.size()};
+    shape = {};
```

Comment:
could this be set to `ArrayRef(storage)` instead of changing the getters?
```suggestion
    shape = ArrayRef(storage);
```

Reply:
No, that's the bug. If you copy the builder and the first is destroyed you end up with array refs pointing at junk. The new getters prevent that.

Reply:
The comment here is saying that you can maintain integrity by pointing to the new storage here instead of clearing the shape and detecting this in the getter.

Reply:
Wouldn't two successive calls to `dropDim` trigger the assertion here?

Reply:
The bug is relying on an `ArrayRef` that when the builder is copied (via the default C++ copy constructor), points to the original builder's storage, rather than the new builder's.  `dropDim()` returns a reference to the builder, so there's no copy done for a chain of them (which makes this bug subtle).

Bug not triggered here:
```
// One builder is constructed. Updated by reference via the dropDim calls.
// Then converted to a vector type.
VectorType newType =  VectorType::Builder(type).dropDim(0).dropDim(1);
```

Bug triggered here:
```
// One a builder is constructed, updated by reference, then assigned 
// to auto, which ends up copying the builder. The original temporary
// builder is destroyed, and `shape` now points to junk.
auto newType =  VectorType::Builder(type).dropDim(0).dropDim(1);
VectorType newVectorType = VectorType(newType);
```

(which is why the CI fix was just: https://github.com/llvm/llvm-project/commit/b44b3494f60296db6aca38a14cab061d9b747a0a)

Reply:
I get that, but I'm not sure how that answers my question right now?

You're setting `shape` to an empty array, but `dropDim` starts with `assert(pos < shape.size() && "overflow");`
So how would `auto newType = VectorType::Builder(type).dropDim(0).dropDim(1);` work?

Seems to me like the second `dropDim(1)` would have `shape.size() == 0` and the assert firing?

Reply:
Oh sorry, I didn't notice that assert. That should be updated to `getShape()`, I would guess it's not triggered right now as noting does multiple `.dropDims()` in a row. The invalid shape bug still exists with a single `dropDim()` call.

Reply:
I think it would be good to have some tests for this, perhaps in `mlir/unittests/IR/ShapedTypeTest.cpp`

Reply:
I've added tests that should catch all the issues discussed in the PR (when run under sanitizers, which I believe does happen).

Reply:
>  That should be updated to getShape(), 

Right, and we're getting to what was asked here: you're making it so that any direct use of `shape` inside this class is "dangerous" now: basically all the code in the implementation should use `getShape()`.

And the question was about instead making sure that the "shape" is always consistent.


That said, I'm question the soundness of this `shape` field entirely: seems like copying the builder would actually drop the field? Also `dropDim` here is reseting the shape field, which seems like unexpected to me: we're just dropping information on the floor??

Reply:
I don't know what you mean by dropping information on the floor. Nothing is lost when the shape is made empty as it's been copied to the storage. 

Anyway, I think the storage filed is a premature optimization. Most shapes will fit in the inline storage of the SmallVector and the main use of the builder (altering dims) requires the storage anyway. So I'll just make shape a SmallVector and remove the copy-on-write stuff.

Reply:
Alternatively, if the COW features are still desired this pattern could be wrapped in a safer class (like  `CopyOnWriteArrayRef<T>`) and that used in the builders instead.

Reply:
> And the question was about instead making sure that the "shape" is always consistent.

As in consistent with `storage`? Do we need that? With this change (*):
* if `shape` **is not** empty then it's consistent with `storage`,
* if `shape` **is** empty then it's inconsistent with `storage` and one should be usiing the latter instead.

This would makes sense to me, but IMHO requires a bit more documentation. For example, you could decorate `shape` with a comment hinting that it shouldn't be used directly.

> I think it'll just wrap this pattern in a helper CopyOnWriteArrayRef<T> class, which can be safer, and simplify the builder's code (and ensure consistency there).

It feels that that's adding complexity to something that was meant to be fairly simple. 

> we're just dropping information on the floor??

The information would be preserved in `storage`, so not really?

Anyway, mostly my 2p. @joker-eph will have a much better intuition what the design goals should be here.

(*) Apologies if I am stating the obvious, but want to make sure that I follow the logic here.

Reply:
>> I think it'll just wrap this pattern in a helper CopyOnWriteArrayRef class, which can be safer, and simplify the builder's code (and ensure consistency there).

> It feels that that's adding complexity to something that was meant to be fairly simple.

It seemed like there was confusion with mixing the logic in the builders, so moving it to a helper class `CopyOnWriteArrayRef<T>`  actually simplifies the builders (and alleviates the concerns about needing the `getShape()/getScalableDims()` methods).

Reply:
> if shape is not empty then it's consistent with storage,
> if shape is empty then it's inconsistent with storage and one should be using the latter instead.

This statement as-is is just leading me to question: what is the purpose of the "shape" field in the first place? As described here it is just redundant with the storage.

Reply:
The builders are copy-on-write (this PR does not change that). The shape is only copied (from the input `ArrayRef`, a pointer to some other storage) on modification (of the shape).

Reply:
> if shape is not empty then storage should be empty.
> if shape is empty then storage contains the information.

Yep, the two fields are mutually exclusive :)